torch/nn/utils/meta_init.py

```diff
@@ -0,0 +1,72 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from contextlib import contextmanager
+from itertools import chain
+from typing import Iterator, Union
+
+import torch
+from torch import Tensor
+from torch.nn import Module
+
+
+@contextmanager
+def meta_init() -> Iterator[None]:
+    """Constructs all tensors within its scope using the meta device.
+
+    This context manager is meant to be used if a module is too expensive or too
+    big to construct on a single machine. It forces all tensors within its scope
+    to use the meta device regardless of their real device. This way all modules
+    constructed inside the context will be allocated on the meta backend.
+
+    Such modules can be used for inspection purposes and then later materialized
+    by calling the :func:`materialize` function.
+    """
+    torch._C._enable_meta_init(True)
+    try:
+        yield
+    finally:
+        torch._C._enable_meta_init(False)
+
+
+def is_meta_init() -> bool:
+    """Indicates whether the caller is within a meta-init context."""
+    return torch._C._is_meta_init_enabled()
+
+
+def materialize(obj: Union[Tensor, Module], keep_cache: bool = False) -> None:
```

Comment:
What is your plan for implementing `materialize`? Are you planning on implementing some lightweight tracing mechanism in C++ that records calls on the objects?

Reply:
Yes, that will be part of the 4th PR. The fallback kernel builds a pretty lightweight unidirected graph of operator calls where the nodes contain the necessary information to "re-construct" the output tensor (e.g. its `Stack`, `ThreadLocalState`, `OperatorHandle`, an "alias id" to determine in-place and view outputs)

Reply:
Probably would be easier to review this if all the materialize tensor changes are in the 4th PR in one place.

Reply:
We should probably also think about how we would materialize ShardedTensor as well. For example, we could initialize an nn.Module on meta device and then use the shard_paramter API: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharded_tensor/__init__.py#L384 to shard a tensor. Now, we need a way to materialize this ShardedTensor.